pkg/expansion/db.go

```diff
@@ -0,0 +1,291 @@
+
+	// Add edges
+	edges := d.edgesForTemplate(template)
+	cycle := false
+	for _, e := range edges {
+		from := hashID(e.x)
+		to := hashID(e.y)
+
+		createsCycle, err := graph.CreatesCycle(d.graph, from, to)
+		if err != nil {
+			return false, fmt.Errorf("checking cycle for template %s: %w", id, err)
+		}
+		cycle = createsCycle || cycle
+
+		if err = d.graph.AddEdge(from, to); err != nil {
+			return false, fmt.Errorf("adding edge for template %s: %w", id, err)
+		}
+	}
+
+	return cycle, nil
+}
+
+func (d *db) edgesForTemplate(template *expansionunversioned.ExpansionTemplate) []edge {
+	var edges []edge
+	id := keyForTemplate(template)
+	genGVK := genGVKToSchemaGVK(template.Spec.GeneratedGVK)
+
+	// Add out-bound edges (from this template's generated GVK to other
+	// templates' matched GVKs)
+	for t := range d.matchers[genGVK] {
+		edges = append(edges, edge{id, t})
+	}
+
+	// Add in-bound edges (from other templates' generated GVK to this template's
+	// matched GVK)
+	for _, gen := range applyToGVKs(template) {
+		for t := range d.generators[gen] {
+			edges = append(edges, edge{t, id})
+		}
+	}
+
+	return edges
+}
+
+func (d *db) handleRemove(template *expansionunversioned.ExpansionTemplate) {
+	id := keyForTemplate(template)
+
+	// The template must exist. Existence checks should be done upstream.
+	if _, exists := d.store[id]; !exists {
+		panic(fmt.Errorf("called handleRemove for template %q, but template DNE in store", id))
+	}
+
+	// Update storage
+	delete(d.store, id)
+
+	// Update generators
+	genGVK := genGVKToSchemaGVK(template.Spec.GeneratedGVK)
+	if _, exists := d.generators[genGVK]; exists {
+		delete(d.generators[genGVK], id)
+		if len(d.generators[genGVK]) == 0 {
+			delete(d.generators, genGVK)
+		}
+	} else {
+		panic(fmt.Errorf("[template %q] inconsistent db - expected key %s to exist in generators", id, genGVK))
+	}
+
+	// Update matchers
+	matches := applyToGVKs(template)
+	for _, m := range matches {
+		if _, exists := d.matchers[m]; exists {
+			delete(d.matchers[m], id)
+			if len(d.matchers[m]) == 0 {
+				delete(d.matchers, m)
+			}
+		} else {
+			panic(fmt.Errorf("[template %q] inconsistent db - expected key %s to exist in matches", id, m))
+		}
+	}
+
+	// Remove edges
+	edges := d.edgesForTemplate(template)
+	for _, e := range edges {
+		from := hashID(e.x)
+		to := hashID(e.y)
+		if err := d.graph.RemoveEdge(from, to); err != nil {
+			panic(fmt.Errorf("[template %q] unexpected error removing edge: %w", id, err))
+		}
+	}
+}
+
+func (d *db) updateCycles() {
+	// First reset all conflicts
+	for _, ts := range d.store {
+		ts.hasConflicts = false
+	}
+
+	conflicts, err := graph.StronglyConnectedComponents(d.graph)
+	if err != nil {
+		panic(fmt.Errorf("error getting SCCs: %w", err))
+	}
+	// All strongly connect components containing more than 1 vertex are a cycle
+	for _, scc := range conflicts {
+		if len(scc) <= 1 {
+			continue
+		}
+		for _, id := range scc {
+			d.store[TemplateID(id)].hasConflicts = true
```

Comment:
Potential performance improvement: storing the set of templates with conflicts as its own map

Reply:
Unless I'm missing something, this would only save us from having to do this loop in `updateCycles` when we reset the templates:
```
for _, ts := range d.store {
	ts.hasConflicts = false
}
```

Obviously this is O(# templates), but each operation is so cheap that I don't think it's worth the code complexity of managing 2 maps.

`add`, `remove`, and `templateFromGVK` would still have the same performance characteristics with the separate maps.

Reply:
Yeah, probably premature.